client/components/web-preview/content.jsx

```diff
@@ -0,0 +1,309 @@
+		}
+		if ( this.props.previewMarkup ) {
+			this.setIframeMarkup( this.props.previewMarkup );
+		}
+
+		if ( typeof this.props.onDeviceUpdate === 'function' ) {
+			this.props.onDeviceUpdate( this.state.device );
+		}
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'message', this.handleMessage );
+	}
+
+	shouldComponentUpdate( nextProps, nextState ) {
+		return shallowCompare( this, nextProps, nextState );
+	}
+
+	componentDidUpdate( prevProps ) {
+		const { previewUrl } = this.props;
+
+		this.setIframeUrl( previewUrl );
+
+		// If the previewMarkup changes, re-render the iframe contents
+		if ( this.props.previewMarkup && this.props.previewMarkup !== prevProps.previewMarkup ) {
+			this.setIframeMarkup( this.props.previewMarkup );
+		}
+		// If the previewMarkup is erased, remove the iframe contents
+		if ( ! this.props.previewMarkup && prevProps.previewMarkup ) {
+			debug( 'removing iframe contents' );
+			this.setIframeMarkup( '' );
+		}
+		// Focus preview when showing modal
+		if ( this.props.showPreview && ! prevProps.showPreview && this.state.loaded ) {
+			this.focusIfNeeded();
+		}
+	}
+
+	handleMessage = ( e ) => {
+		let data;
+		try {
+			data = JSON.parse( e.data );
+		} catch ( err ) {
+			return;
+		}
+
+		if ( ! data || data.channel !== 'preview-' + this.previewId ) {
+			return;
+		}
+		debug( 'message from iframe', data );
+
+		switch ( data.type ) {
+			case 'link':
+				page( data.payload.replace( /^https:\/\/wordpress\.com\//i, '/' ) );
+				if ( typeof this.props.onClose === 'function' ) {
+					this.props.onClose();
+				}
+				return;
+			case 'close':
+				if ( typeof this.props.onClose === 'function' ) {
+					this.props.onClose();
+				}
+				return;
+			case 'partially-loaded':
+				this.setLoaded();
+				return;
+		}
+	}
+
+	focusIfNeeded = () => {
+		// focus content unless we are running in closed modal or on empty page
+		if ( this.iframe.contentWindow && this.state.iframeUrl !== 'about:blank' ) {
+			debug( 'focusing iframe contents' );
+			this.iframe.contentWindow.focus();
+		}
+	}
+
+	setIframeMarkup( content ) {
+		if ( ! this.iframe ) {
+			debug( 'no iframe to update' );
+			return;
+		}
+		debug( 'adding markup to iframe', content.length );
+		this.iframe.contentDocument.open();
+		this.iframe.contentDocument.write( content );
+		this.iframe.contentDocument.close();
+	}
+
+	setIframeUrl = ( iframeUrl ) => {
+		if ( ! this.iframe || ( ! this.props.showPreview && this.props.isModalWindow ) ) {
+			return;
+		}
+
+		// Bail if we've already set this url
+		if ( iframeUrl === this.state.iframeUrl ) {
+			return;
+		}
+
+		debug( 'setIframeUrl', iframeUrl );
+		try {
+			const newUrl = iframeUrl === 'about:blank'
+				? iframeUrl
+				: addQueryArgs( { calypso_token: this.previewId }, iframeUrl );
+			this.iframe.contentWindow.location.replace( newUrl );
+			this.setState( {
+				loaded: false,
+				iframeUrl: iframeUrl,
+			} );
+		} catch ( e ) {}
```

Comment:
I was going to comment the same about this try block. Let's see if we can't pull it out or narrow the scope in a follow-up PR.